Update SpeedGrader landscape layout

vargaat opened the pull request:
### What's new?
- Updated landscape layout according to the new design.
- Created a `onSizeChange` helper to measure views. This is more effective then previously used methods. We could even use this instead of the lately introduced `measuringSize` methods.
- The main content area is now resizable by dragging the new collapse icon. Drag position is persisted when swiping between students.
- Tapping the dragger icon will show / hide the right drawer. The show action will restore the last extended position.
- Introduced a new segmented picker. Marci wants to use this iOS component so I added it to InstUI and renamed the previous one that is only used in SpeedGrader and can be deleted after we update the portrait layout. Customizing the appearance of this native component is not straightforward so we might need a follow-up ticket to tackle any issues with it. We might need to re-write it from scratch just as we did with the switch.

refs: [MBL-18832](https://instructure.atlassian.net/browse/MBL-18832)
affects: Teacher
release note: SpeedGrader's drawer size is now resizable in landscape mode.

test plan:
- See ticket for expectations.

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td><img src="https://github.com/user-attachments/assets/d60fcc10-509f-4445-a730-e30f271fcd4f" maxHeight=500></td>
<td><img src="https://github.com/user-attachments/assets/6c4b1814-5372-4cfa-8e44-1347f575b7d3" maxHeight=500></td>
</tr>
<tr>
<td></td>
<td><img src="https://github.com/user-attachments/assets/44c7caeb-d246-4a24-abbb-7f46c3ba3fea" maxHeight=500></td>
</tr>
<tr>
<td></td>
<td><img src="https://github.com/user-attachments/assets/c2fa98ae-2bf3-4467-a318-e9bf57f2efdb" maxHeight=500></td>
</tr>
</table>

## Checklist

- [ ] Follow-up e2e test ticket created
- [x] A11y checked
- [ ] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode
- [x] Approve from product


[MBL-18832]: https://instructure.atlassian.net/browse/MBL-18832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

inst-danger commented:
<!--
  2 failure:  Build failed, ski..., Please run `yarn ...
  0 warning: 
  
  3 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please run `yarn update-headers`. The following do not have the correct license header: 
* Teacher/TeacherTests/SpeedGrader/MainLayout/ViewModel/SpeedGraderLandscapeSplitLayoutViewModelTests.swift
  </td>
    </tr>
  </tbody>
</table>



#### Release Note: 
SpeedGrader's drawer size is now resizable in landscape mode.

#### Affected Apps: Teacher

[MBL-18832](https://instructure.atlassian.net/browse/MBL-18832)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 240e45e039dd392029f01dd0cd3b850641fa74a7
</p>

inst-danger commented:
<details><summary>Teacher Build QR Code:</summary>
<img src="https://api.qrserver.com/v1/create-qr-code/?data=https%3A%2F%2Fapp.bitrise.io%2Fapp%2F88a3641e29e4745c%2Finstallable-artifacts%2F6d488879f531ca6e%2Fpublic-install-page%2Fe2e8eb92ed08a8c2b2fb103e9d019e1f&size=150x150" />
</details>


<!-- TeacherBuild -->

suhaibabsi-inst commented on Teacher/Teacher/SpeedGrader/MainLayout/View/SubmissionGraderView.swift:
```diff
@@ -133,13 +149,14 @@ struct SubmissionGraderView: View {
             }
+            .frame(width: landscapeSplitLayoutViewModel.leftColumnWidth)
+            InstUI.Divider()
+            VStack(spacing: 0) {
```
Let's move the `VStack` to the `tools(...)` method.

suhaibabsi-inst commented on Teacher/Teacher/SpeedGrader/MainLayout/View/SubmissionHeaderView.swift:
```diff
@@ -58,9 +61,15 @@ struct SubmissionHeaderView: View {
             })
-                .buttonStyle(PlainButtonStyle())
-                .identifier("SpeedGrader.userButton")
+            .buttonStyle(PlainButtonStyle())
```
```suggestion
            .buttonStyle(.plain)
```

suhaibabsi-inst reviewed (commented):
Code +1

suhaibabsi-inst commented on Core/Core/Common/CommonUI/SwiftUIViews/View+MeasuringSize.swift:
```diff
@@ -50,4 +50,20 @@ extension View {
     }
+
+    public func onSizeChange(_ perform: @escaping (CGSize) -> Void) -> some View {
+        onGeometryChange(for: CGSize.self) { geometry in
```
[This](https://developer.apple.com/documentation/swiftui/view/ongeometrychange(for:of:action:)) is interesting!.
It is good that they addressed such needs.
There is also [this one](https://developer.apple.com/documentation/swiftui/view/onscrollgeometrychange(for:of:action:)) that can be useful to track scroll offset.

suhaibabsi-inst reviewed (approved):
QA +1 

Tested on iPad Simulator.

One note though, when you hide the right drawer on Landscape, then switch to portrait then switch back to Landscape, the right drawer shows again. See recording below

https://github.com/user-attachments/assets/60308616-791f-4235-abc3-f0442709386c

vargaat commented:
> One note though, when you hide the right drawer on Landscape, then switch to portrait then switch back to Landscape, the right drawer shows again. See recording below

Nice catch. The layout was also reset if the app was put into background. I fixed both of these.

suhaibabsi-inst reviewed (approved):
QA +1

vargaat commented on Teacher/Teacher/SpeedGrader/StudentPager/Model/SpeedGraderInteractorLive.swift:
```diff
@@ -58,7 +58,7 @@ class SpeedGraderInteractorLive: SpeedGraderInteractor {
             SpeedGraderContextInfo(
                 courseName: course.name ?? "",
-                courseColor: course.color,
+                courseColor: Color(course.color),
```
Done.

vargaat commented on Teacher/Teacher/SpeedGrader/MainLayout/View/SubmissionGraderView.swift:
```diff
@@ -230,79 +254,105 @@ struct SubmissionGraderView: View {
     }
 
     @ViewBuilder
     private func tools(bottomInset: CGFloat, isDrawer: Bool) -> some View {
```
Agree, we'll have opportunities to do this while working on the SpeedGrader epic.

vargaat commented on Teacher/Teacher/SpeedGrader/MainLayout/ViewModel/SpeedGraderLandscapeSplitLayoutViewModel.swift:
```diff
@@ -0,0 +1,132 @@
+    @Published private(set) var isRightColumnHidden = false
+    @Published private(set) var rightColumnWidth: CGFloat?
+    @Published private(set) var leftColumnWidth: CGFloat? {
+        didSet {
```
Moved.

vargaat commented on Teacher/Teacher/SpeedGrader/MainLayout/ViewModel/SpeedGraderLandscapeSplitLayoutViewModel.swift:
```diff
@@ -0,0 +1,132 @@
+    @Published private(set) var rightColumnWidth: CGFloat?
+    @Published private(set) var leftColumnWidth: CGFloat? {
+        didSet {
+            let isRightColumnHidden = (leftColumnWidth == screenWidth)
+            withAnimation(.snappy) {
```
Good suggestion.

vargaat commented on Teacher/Teacher/SpeedGrader/MainLayout/ViewModel/SpeedGraderLandscapeSplitLayoutViewModel.swift:
```diff
@@ -0,0 +1,132 @@
+    func didTapDragIcon() {
+        if isFullScreen {
+            withAnimation(.snappy) {
+                let leftColumnWidth = leftColumnWidthBeforeFullScreen ?? leftSideMaxWidth
```
The block here is not escaping and will run synchronously so we don't need to capture anything in it.

vargaat commented on Teacher/Teacher/SpeedGrader/MainLayout/ViewModel/SpeedGraderLandscapeSplitLayoutViewModel.swift:
```diff
@@ -0,0 +1,132 @@
+            leftColumnWidthBeforeFullScreen = nil
+        } else {
+            leftColumnWidthBeforeFullScreen = leftColumnWidth
+            withAnimation(.snappy) {
```
Done.

vargaat commented:
> As an alternative would it make sense to handle it similarly to Profile Menu button?

I did this in the end.

vargaat commented:
> Would it make sense to focus the selected segment tab upon opening the drawer?

Done, it was an easy addition.

vargaat commented:
> SegmentedPicker:
> - The only issue I found that the font doesn't scale dynamically (but it scales if the font size is already large before opening the screen), not even if dynamicTypeSize is used.
> - This may or may not be what we want, but it's a minor thing IMHO.

This is known. I played around for a few hours on how to properly customize this view but with the appearance selectors it's not that easy so I just left it as-is.

> If the device is rotated to portrait while a drag is in progress it seems to save the mid-flight pane width value, which may be invalid (= lower then the minimum). Rotating back to landscape sets this unvalidated pane width. Tapping the button in this state also just toggles the pane without validation.
> I don't think users would rotate the device on purpose while dragging, but it's not hard for this to happen accidentally (that's what happened to me, as well)

Fixed.

> Content resizing (via drag or tap):
> - Some submission contents (like webviews) react badly to the continuous resizing while dragging or animating. The UI becomes unresponsive (almost frozen in some cases). Maybe we should avoid constant resizing in some (or all?) cases.
> (example: my account, "All Submission Types", Kid 1, Attempt 7 & 8)

I improved the drag performance by not using animations during the interactive drag.

> The bounce animation is not be ideal here, IMHO. Maybe smooth would be better to avoid any bouncing. (especially because it continuously resizes the submission content, too)

Good call, I changed the animation and made it faster.

Other performance problems: I suggest to revisit this topic after most of the redesign efforts are completed. For example when we have a lot of comments it's really slow to format dates during each render cycle on the fly, these formatted dates could be cached in a dedicated view model for the cell but that's just one thing we could improve.

rh12 reviewed (approved):
QA + 1 (tested on iPad @ 17)

Thanks for the changes! It's much more responsive now!
